### PR TITLE
Evaluate symbolic links on Add/Copy Commands

### PR DESCRIPTION
**Pull request by rhatdan**
Users expect to be able to use Symbolic links in the
ADD and COPY commands in Dockerfiles.  We need to evaluate
these and pull in the correct content.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Fixes https://github.com/projectatomic/buildah/issues/1006
@TomSweeneyRedHat @nalind @pixdrift @vrothberg PTAL

**Comment by rhatdan** [comment]
https://github.com/Apkubatrh PTAL

**Comment by rhatdan** [comment]
@umohnani8 @boaz1337 PTAL

File: tests/copy.bats [hunk 1]
```diff
@@ -139,3 +139,22 @@ load helpers
...
+  root=$(buildah mount $cid)
+  buildah config --workingdir / $cid
+  buildah copy $cid ${TESTDIR}/link-randomfile
+  buildah unmount $cid
+  buildah commit --signature-policy ${TESTSDIR}/policy.json $cid containers-storage:new-image
+  buildah rm $cid
+
+  newcid=$(buildah from --signature-policy ${TESTSDIR}/policy.json new-image)
+  newroot=$(buildah mount $newcid)
+  test -s $newroot/randomfile
```
Comment (nalind) [inline tests/copy.bats]: Shouldn't we expect `link-randomfile` to keep its name when it's copied?

Comment (rhatdan) [reply tests/copy.bats]: I don't know, I will check with Docker.

**Comment by rhatdan** [comment]
@nalind PTAL

File: add.go [hunk 2]
```diff
@@ -168,10 +168,14 @@ func (b *Builder) Add(destination string, extract bool, options AddAndCopyOption
 			return errors.Wrapf(syscall.ENOENT, "no files found matching %q", src)
 		}
 		for _, gsrc := range glob {
-			srcfi, err := os.Stat(gsrc)
+			esrc, err := filepath.EvalSymlinks(gsrc)
 			if err != nil {
 				return errors.Wrapf(err, "error reading %q", gsrc)
```
Comment (nalind) [inline add.go]: Please tweak this error message so that we can distinguish it from the next one.

File: tests/copy.bats [hunk 3]
```diff
@@ -139,3 +139,22 @@ load helpers
...
+  root=$(buildah mount $cid)
+  buildah config --workingdir / $cid
+  buildah copy $cid ${TESTDIR}/link-randomfile
+  buildah unmount $cid
+  buildah commit --signature-policy ${TESTSDIR}/policy.json $cid containers-storage:new-image
+  buildah rm $cid
+
+  newcid=$(buildah from --signature-policy ${TESTSDIR}/policy.json new-image)
+  newroot=$(buildah mount $newcid)
+  test -s $newroot/link-randomfile
```
Comment (nalind) [inline tests/copy.bats]: Please also ensure that the new `link-randomfile` isn't a symbolic link (using `test -h`).

Comment (rhatdan) [reply tests/copy.bats]: Added test -f

Comment (nalind) [reply tests/copy.bats]: That looks better than what I was thinking, actually.

**Comment by nalind** [comment]
LGTM

**Comment by umohnani8** [comment]
LGTM

**Comment by rhatdan** [comment]
@rh-atomic-bot r=nalind

**Pull request closed without merging**